DEVINFO.md

```diff
@@ -0,0 +1,174 @@
+
+The size and scale of a PR depend on the reviewers' abilities to process the change.  Different reviewers and submitters have different levels of familiarity with the codebase.  Moreover, everyone has a different schedule -- sometimes, some people are more busy than others.  A successful PR submitter **takes the reviewers' familiarity and availability into account** when crafting the PR, even going to far as to ask in advance if a particular person could be available for review.
+
+## Submission Checklist
+
+A PR submission's text should **answer the following questions** for reviewers:
+
+* What problem is being solved by this PR?
+* What does the solution do to address them?
+* Why is this the best solution?  What alternatives were considered, and why are they worse?
+* What do reviewers need to be familiar with in order to provide useful feedback?
+* What issue(s) are addressed by this PR?
+* What are some hints to understanding some of the more intricate or clever parts of the PR?
+
+In addition, the PR submission should **answer the prompts of the Github template** we use for PRs.
+
+The code itself should adhere to the following guidelines, which both submitters and reviewers should check:
+
+### Documentation
+
+* Each file must have a **copyright statement**.
+* Any new non-test modules should have **module-level documentation** explaining what the module does, and how it fits into the blockchain as a whole.
+* Any new files must have some **top-of-file documentation** that describes what the contained code does, and how it fits into the overall module.
+
+Within the source files, the following **code documentation** standards are expected:
+
+* Each public function, struct, enum, and trait should have a Rustdoc comment block describing the API contract it offers.  This goes for private structs and traits as well.
+* Each _non-trivial_ private function should likewise have a Rustdoc comment block.  Trivial ones that are self-explanatory, like getters and setters, do not need documentation.  If you are unsure if your function needs a docstring, err on the side of documenting it.
+* Each struct and enum member must have a Rustdoc comment string indicating what it does, and how it is used.  This can be as little as a one-liner, as long as the relevant information is communicated.
+
+### Factoring
+
+* **Public or exported struct, enum, and trait definitions go into the `mod.rs` file**.  Private structs, enums, and traits can go anywhere.
```

Comment:
This isn't a settled convention in our codebase, and I'd for one argue against this -- putting struct defs in `mod.rs` files and implementations elsewhere often makes finding method definitions very confusing (without the assistance of rust-analyzer or likewise).

Reply:
I'll second Aaron's comment. I'm still new to the codebase and I'm noticing we have a lot in our `mod.rs` files. For me, it's making things harder to find than if all code for related to a subsystem was in the same file with a descriptive name. The convention I've used before was to use `mod.rs` only for things that were common to many of the files in the module.

Reply:
I don't think that's an excuse to start developing good habits, though.  The best time to do it was when we started; the second-best time is today.

I'm not too tied to this _particular_ convention, but I think we need to define a _reasonable_ convention that we can stick to.  Right now, for example, `clarity/` is organized _very_ differently than `src/chainstate`.

Reply:
I think we should settle on something, but I don't think it should block this PR. I think that given the current state of contributing docs in this repo, we should opt to merge this PR pretty soon, and then refinements can always be made in subsequent PRs.

I opened an issue for this (#3641) and I'm going to remove this line from the PR.

Reply:
Removed in 2b5cda5b6